### PR TITLE
fix group symbols cached bug

### PR DESCRIPTION
**Pull request by ZihengJiang**
- change the interface of Session.Run(...)
- use a hash function to generate the hash value of the vector of Symbols*

### CODE REVIEW

File: src/session.cc [hunk 1]
```diff
@@ -144,11 +144,28 @@ Session* Session::Create(const std::string& option) {
...
+  for (uint32_t i = 0; i < syms.size(); ++i) {
+    sym_arr.push_back(*syms[i]);
+  }
+  *sym = nnvm::Symbol::CreateGroup(sym_arr);
+  uint64_t hash_value = hash_symbols(syms);
+
+  if (cached_execs_.count(hash_value) != 0) {
+    auto& entry = cached_execs_.at(hash_value);
     const nnvm::Symbol& s = entry.exec->symbol();
     bool stale_exec = (s.outputs.size() != sym->outputs.size());
```
Comment (tqchen) [inline src/session.cc]: This check is not enough, need to check the ptr of all symbols

Comment (ZihengJiang) [reply src/session.cc]: @tqchen looks like this requires us to save the pointer of symbols before. Do you mean there is a condition that two vector of symbols have the same hash value and also can pass the stale executor check below?

Comment (tqchen) [reply src/session.cc]: Yes. thatiswhat I mean

File: src/session.cc [hunk 2]
```diff
@@ -145,35 +147,65 @@ Session* Session::Create(const std::string& option) {
...
-  if (cached_execs_.count(sym) != 0) {
-    auto& entry = cached_execs_.at(sym);
-    const nnvm::Symbol& s = entry.exec->symbol();
-    bool stale_exec = (s.outputs.size() != sym->outputs.size());
+  nnvm::Symbol* new_sym = new nnvm::Symbol();
+  std::vector<nnvm::Symbol> sym_arr;
+  for (uint32_t i = 0; i < syms.size(); ++i) {
+    sym_arr.push_back(*syms[i]);
+  }
+  *new_sym = nnvm::Symbol::CreateGroup(sym_arr);
```
Comment (tqchen) [inline src/session.cc]: simply write nnvm::Symbol new_sym instead of pointer

File: src/session.cc [hunk 3]
```diff
@@ -145,35 +147,65 @@ Session* Session::Create(const std::string& option) {
...
+
+  // compute the hash value of the input symbols
+  uint64_t hash_value = syms.size();
+  for (nnvm::Symbol* i : syms) {
```
Comment (tqchen) [inline src/session.cc]: Let us add a new function called FindCachedExecs

File: src/session.cc [hunk 4]
```diff
@@ -144,55 +146,66 @@ Session* Session::Create(const std::string& option) {
...
   for (uint32_t i = 0; i < syms.size(); ++i) {
     sym_arr.push_back(*syms[i]);
   }
-  *sym = nnvm::Symbol::CreateGroup(sym_arr);
-  uint64_t hash_value = hash_symbols(syms);
+  *new_sym = nnvm::Symbol::CreateGroup(sym_arr);
+
+  // compute the hash value of the input symbols
+  uint64_t hash_value = syms.size();
+  for (nnvm::Symbol* i : syms) {
```
Comment (tqchen) [inline src/session.cc]: Use pointer of nodes as hash keys, this removes the need of symbol pointers

File: src/session.cc [hunk 5]
```diff
@@ -67,6 +67,8 @@ class TorchSession : public Session {
  private:
   // entry to store cached executor
   struct ExecEntry {
+    // cached symbols
+    std::vector<nnvm::Symbol*> cached_symbols_;
```
Comment (tqchen) [inline src/session.cc]: Simply use nnvm::Symbol, use the node pointers as key

Comment (ZihengJiang) [reply src/session.cc]: do you mean group symbol? then use the pointer of output node?

Comment (ZihengJiang) [reply src/session.cc]: the output node will change, so it should be `symbol.outputs[0].inputs` nodes?

Comment (tqchen) [reply src/session.cc]: the output node won't change, as they are refering back to the same NodePtr

**Comment by ZihengJiang** [comment]
@tqchen 
- use pointer of nodes as hash value
- I recommend to add a cached symbol in ExecEntry, due to the exec.symbol() maybe will be changed after we change the graph

**Pull request merged**